src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,682 @@
+    /// the underlying stream into a single operation. This is because we assume that memory copies are significantly
+    /// faster than IO operations on the underlying stream (if this was not true, using buffering is never appropriate).
+    /// The max size of this "shadow" buffer is limited as to not allocate it on the LOH.
+    /// Shadowing is always transient. Even when using this technique, this class still guarantees that the number of
+    /// bytes cached (not yet written to the target stream or not yet consumed by the user) is never larger than the 
+    /// actual specified buffer size.
+    /// </summary>
+    [ComVisible(true)]
+    public sealed partial class BufferedStream : Stream
+    {
+        private const Int32 _DefaultBufferSize = 4096;
+
+
+        private Stream _stream;                               // Underlying stream.  Close sets _stream to null.
+
+        private Byte[] _buffer;                               // Shared read/write buffer.  Alloc on first use.
+
+        private readonly Int32 _bufferSize;                   // Length of internal buffer (not counting the shadow buffer).
+
+        private Int32 _readPos;                               // Read pointer within shared buffer.
+        private Int32 _readLen;                               // Number of bytes read in buffer from _stream.
+        private Int32 _writePos;                              // Write pointer within shared buffer.
+
+        /// <summary><code>MaxShadowBufferSize</code> is chosed such that shadow buffers are not allocated on the Large Object Heap.
+        /// Currently, an object is allocated on the LOH if it is larger than 85000 bytes. See LARGE_OBJECT_SIZE in ndp\clr\src\vm\gc.h
+        /// We will go with exactly 80 KBytes, although this is somewhat arbitrary.</summary>
+        private const Int32 MaxShadowBufferSize = 81920;  // Make sure not to get to the Large Object Heap.
+
+        // Removing a private default constructor is a breaking change for the DataContractSerializer.
+        // Because this ctor was here previously we need to keep it around.
+        private BufferedStream() { }
+
+        public BufferedStream(Stream stream)
+
+            : this(stream, _DefaultBufferSize)
+        {
+        }
+
+        public BufferedStream(Stream stream, Int32 bufferSize)
+        {
+            if (stream == null)
+                throw new ArgumentNullException("stream");
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException("bufferSize", SR.Format(SR.ArgumentOutOfRange_MustBePositive, "bufferSize"));
+
+            Contract.EndContractBlock();
+
+            Contract.Assert(!(stream is MemoryStream), "MemoryStream shouldn't be wrapped in a BufferedStream!");
+            Contract.Assert(!(stream is BufferedStream), "BufferedStream shouldn't be wrapped in another BufferedStream!");
```

Comment:
All of the Contract.Asserts and Contract.Requires should be changed to Debug.Asserts.  The Contract.EndContractBlocks can be removed.